persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/ClinicalDataSparkRepository.java

```diff
@@ -0,0 +1,280 @@
+    @Override
+    public List<ClinicalData> fetchClinicalData(List<String> studyIds, List<String> sampleIds, List<String> attributeIds,
+                                                String clinicalDataType, String projection) {
+        List<Dataset<Row>> res = new ArrayList<>();
+        for (String studyId : new HashSet<>(studyIds)) {
+            spark.sqlContext().setConf("spark.sql.caseSensitive", "false");
+            String dataFile = ParquetConstants.CLINICAL_PATIENT;
+            if (PersistenceConstants.SAMPLE_CLINICAL_DATA_TYPE.equalsIgnoreCase(clinicalDataType)) {
+                dataFile = ParquetConstants.CLINICAL_SAMPLE;
+            } 
+            Dataset<Row> clinicalData = spark.read()
+                .option("mergeSchema", true)
+                .parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + dataFile);
+
+            if ("PATIENT".equalsIgnoreCase(clinicalDataType)) {
+                Dataset<Row> clinicalSampleData = spark.read()
+                    .option("mergeSchema", true)
+                    .parquet(PARQUET_DIR + ParquetConstants.STUDIES_DIR + studyId + "/" + ParquetConstants.CLINICAL_SAMPLE);
+                clinicalData = clinicalData.join(clinicalSampleData, "PATIENT_ID");
+            }
+
+            // filter by sample id
+            if (sampleIds != null && !sampleIds.isEmpty()) {
+                String[] sampleArr = sampleIds.stream().toArray(String[]::new);
+                clinicalData = clinicalData.filter(col("PATIENT_ID").isin(sampleArr).or(col("SAMPLE_ID").isin(sampleArr)));
+            }
+
+            // if attributes is null return data from all columns
+            List<String> attributes = attributeIds;
+            if (attributeIds == null || attributeIds.isEmpty()) {
+                attributes = Arrays.asList(clinicalData.columns());
+            }
+            clinicalData.createOrReplaceTempView("clinicalData");
+
+            for (String c : attributes) {
+                StringBuilder sb = new StringBuilder();
+                sb.append("SELECT PATIENT_ID as patientId, SAMPLE_ID as sampleId");
+                if (!MUTATION_COUNT.equalsIgnoreCase(c) && !FRACTION_GENOME_ALTERED.equalsIgnoreCase(c)) {
+                    sb.append(", " + c + " AS attrValue");
+                }
+                sb.append(" FROM clinicalData ");
+
+                String query = sb.toString();
+                Dataset<Row> sub = spark.sql(query);
+
+                sub = sub.withColumn("attrId", lit(c))
+                    .withColumn("studyId", lit(studyId));
+
+                if (MUTATION_COUNT.equalsIgnoreCase(c)) {
+                    sub = sub.join(getMutationCount(studyId), "sampleId");
+                }
+                if (FRACTION_GENOME_ALTERED.equalsIgnoreCase(c)) {
+                    sub = sub.join(getFractionGenomeAltered(studyId), "sampleId");
+                }
+                res.add(sub);
+            }
+        }
+        Dataset<Row> ds = res.get(0);
```

Comment:
could you add comments explaining why this is done. especially line 138

Reply:
Sure I can add the comment. i'm doing this to union all results in datasets before calling collectAsList - which is an expensive call.